### ISSUE TITLE
Can't run any of the commands

### ISSUE THREAD

**Issue by jorgeas80** [buehler/typescript-hero#79]
When I hit Ctrl+alt+g, I get this warning:

> Running the contributed command:'typescriptHero.showCmdGui' failed.

When I hit Ctrl+alt+i, I get this warning:

> Please wait a few seconds longer until the symbol cache has been build.

In both cases, nothing happens after that.

OS: Ubuntu 16.04
VSCode v1.5.3
TypeScript v2.0.3

**Comment by buehler** [comment]
Hey @jorgeas80 
It could be an issue with TypeScript 2.0
Do you see a "flame" in your infobar right next to the "Resolver" ?

If there is a flame, there was an error and you should get an output `TypeScript Hero Extension` where this error would be logged.

Maybe there is an error with ubuntu as well 😢 

Could you check if you had any error outputs?
Cheers

**Comment by jorgeas80** [comment]
Hi @buehler 

I see the flame, but when I click, it just restarts the index process

Check the screenshot: http://pix.toile-libre.org/upload/original/1475508433.png

Running code from command line doesn't help. It doesn't generate any output :-(

**Comment by buehler** [comment]
Alright, could you set the verbosity to "all" and send me the logs?
the setting is `typescriptHero.verbosity`. This can be set to `All`.

Afterwards, have a look at your outputs:
![outputs](https://dl.dropboxusercontent.com/spa/agn3r6rkwi4rxn7/wvyb1v3i.png)

The output window is found at the view menu :-)
If it throws an error, it would be cool to receive those logs..

**Comment by jorgeas80** [comment]
This is what I get. 

![error_log_output](http://pix.toile-libre.org/upload/original/1475510100.png)

I have several angular-cli generated projects as subfolders of the same root folder. Each subfolder (there are 12 of them) have its own node_modules folder (I know, I could just create symlinks). Maybe the extension is trying to index ALL the folders (I have the root folder open in the editor, so, all that node_modules folders are available...)

**Comment by buehler** [comment]
Thanks for the details!
Actually I got an idea what this could be..

The problem with the syntax tree is, that you can create stuff without names. E.g.:
```typescript
export default class {
   /*...*/
}
```

This syntax is totally valid, but horrible to parse. This class's got no name. So the property `text` of the Name-Literal node in the AST is actually hitting an undefined node. This is what booms here.

Do you have any examples open sources? Otherwise, I'll check, which literals aren't checked.

**Comment by jorgeas80** [comment]
Yes, just create a new project with angular-cli:

ng new myapp

Open it with VSCode and hit Ctrl+Alt+g, for example. Or as an alternative, clone this repo https://github.com/jorgeas80/angular-cli-starter (the basic boilerplate created with angular-cli)

**Comment by Chudesnov** [comment]
Another example: https://github.com/eldargab/translit/blob/master/index.js

```js
module.exports = function(map) {
    /*...*/
}
```

exported function has no name, so in typings (.d.ts) for that module in my project I wrote:

```ts
declare module 'translit' {
  /*...*/
  export default function(mapping: IMapping): ITranslitFunction;
}
```

In VSCode extensions console I get this from TS Hero (reduced output):

```
TypeError: Cannot read property 'text' of undefined
   at TsResourceParser.parseFunction 
   (src/parser/TsResourceParser.js)
```

**Comment by buehler** [comment]
Hey @jorgeas80 @Chudesnov 
I fixed the bug and tested it with a plain project of the angular-cli. It was indeed the default exports without a name.
It does work with a plain project and I plan to release 0.9 this weekend.

I'd be happy to hear any feedback if it's not working for you guys.

Cheers

**Comment by jorgeas80** [comment]
Hi @buehler, it's working for me now :-). Thanks for your effort. Kudos for you, man! :+1:

**Comment by gund** [comment]
Hi, I also experiencing this issue however my "Resolve" icon is not burning - it is just "okay".
And here's what I found in the output:

```text
ResolveIndex: ERROR	 Catched an error during buildIndex()
	Data:	TypeError: index[declaration.name].push is not a function
    at ResolveIndex.<anonymous> (C:\Users\oleksii.malkevych\.vscode\extensions\rbbit.typescript-hero-0.11.0\out\src\caches\ResolveIndex.js:332:45)
    at next (<anonymous>)
    at C:\Users\oleksii.malkevych\.vscode\extensions\rbbit.typescript-hero-0.11.0\out\src\caches\ResolveIndex.js:19:65
    at __awaiter (C:\Users\oleksii.malkevych\.vscode\extensions\rbbit.typescript-hero-0.11.0\out\src\caches\ResolveIndex.js:15:12)
    at ResolveIndex.createIndex (C:\Users\oleksii.malkevych\.vscode\extensions\rbbit.typescript-hero-0.11.0\out\src\caches\ResolveIndex.js:311:16)
    at ResolveIndex.<anonymous> (C:\Users\oleksii.malkevych\.vscode\extensions\rbbit.typescript-hero-0.11.0\out\src\caches\ResolveIndex.js:121:42)
    at next (<anonymous>)
    at fulfilled (C:\Users\oleksii.malkevych\.vscode\extensions\rbbit.typescript-hero-0.11.0\out\src\caches\ResolveIndex.js:16:58)
    at process._tickCallback (internal/process/next_tick.js:103:7)
```

Every time I hit Reindex - this error above pops up in the output.
I also tried to reinstall VSCode and remove all other extensions but still having this issue.

Can anyone assist me what it can be?

Thanks.

**Comment by buehler** [comment]
Hey @gund 
This seems to be an issue with the indexing process. Thanks for the error description, I'll have a look into that

**Comment by buehler** [comment]
hey @gund
do you experience this error with certain code ? I cannot find the error :-/
Everything there (where your stacktrace points) is checked for null / undefined code parts

**Comment by gund** [comment]
Hi @buehler,

I'm working on angular2 project with all it's dependencies plus NgRx, NgBootstrap, HMR.
Also I have some @types for TS like angular(v1), jasmine.

I was trying to set a breakpoint just before it fails to see a call stack but was unable to pause there for some reason.

If you know how to debug you can tell me and will try to give you more info about it.

Thanks.

**Comment by buehler** [comment]
Well the easiest way to debug it, would be:
- check out the repository
- do all the installing stuff (npm / yarn / whatever install)
- open the folder with vscode
- `npm run compile`
- debug it with vscode (you can generate the launch.json by vscode for `extension development`)

This way, a new vscode window will open with your debugger attached and a message appears that this code will overwrite your installed extension for this session (exactly what you want)

**Comment by gund** [comment]
Thanks, I will try to look at it and get back to you with result.

**Comment by gund** [comment]
Hi @buehler,

I was able to debug your extension on my project and funny thing is when I open it in vscode with debugger connected there is no exception happening,
However if I just open a project (without debugger) I still see that error in the console.

I even tried to stop on all exceptions but still was not able to catch that,
Very weird situation =)

Do you have any idea how can I try to catch error without debugger?

**Comment by maxisam** [comment]
![image](https://cloud.githubusercontent.com/assets/456807/21057900/1bd72e46-be01-11e6-8343-e17abfac3b73.png)

Here is my error. It is just like @gund 's error

If I exclude node_module folder, it works fine.

**Comment by gund** [comment]
Hey @maxisam, thank you for info, will definitely try it and post back with the result.

**Comment by maxisam** [comment]
@buehler so I get to the bottom of this. It turns out

Here is the line cause the issue. 

`export declare function toString(value: any): string;`

from (util.d.ts)[https://github.com/ng-bootstrap/ng-bootstrap/blob/master/src/util/util.ts]

**Comment by buehler** [comment]
Hey @maxisam 
Hmm thanks, I'll have a look into it :-)

**Comment by noelklein** [comment]
I'm having the exact same problem since I added ng-bootstrap to my angular 2 project. Any way I can help out to get this resolved?

**Comment by gund** [comment]
I added to `.vscode/settings.json` next config:
```
typescriptHero.resolver.ignorePatterns: [
  "node_modules/@ng-bootstrap"
]
```

But still got the error described above.

**Comment by maxisam** [comment]
@gund  That is not what I did. You wanna try my workaround ?

**Comment by gund** [comment]
@maxisam, sure, can you explain where I need to add that line you posted above?
Thanks

**Comment by maxisam** [comment]
It should be in your setting.json. find files.exclude and add that line

**Comment by gund** [comment]
Ok, I will add that line since I hadn't `files.exclude` there before =)

**Comment by gund** [comment]
I've tried what you said but still got same error in console and no commands are executable =(

**Comment by maxisam** [comment]
That is odd. Do you reload the window ?

**Comment by gund** [comment]
I did window reload and also full restart of vscode and nothing changed. Maybe there's also some library that causes it.

**Comment by buehler** [comment]
Mhmm.. as a veeeery (like _veryvery_) dirty hacky workaround, you could just comment out (or delete) the dependency in your `package.json` but leave the package installed. The extension whitelists packages to index via the dependencies or devDepenedencies if there is a package.json file.

**Comment by buehler** [comment]
@noelklein Is already helps that `ng-bootstrap` causes some trouble. If you find the time to debug some app code, that would be very nice, to point into the right direction :-)

**Comment by noelklein** [comment]
maxisam's workaround worked for me, so going with that until after the holidays.

**Comment by buehler** [comment]
I did push a branch with a possible solution: https://github.com/buehler/typescript-hero/tree/fix/ext-crash

I hope this prevents those crashes 😐

**Comment by gund** [comment]
Hi @buehler, thanks for the update.
I will surely test that branch, probably in about 2 days as I will need to access that codebase where it has problems.

Thanks for your work!

**Comment by buehler** [comment]
I've got some ideas, maybe it's just happening when there is a big codebase. I need to rethink this maybe (how the extension generally handles this index)

**Comment by gund** [comment]
Hi @buehler, I was able to run code on your fix branch ans got next Error:
```
ResolveIndex: ERROR	 Catched an error during buildIndex() TypeError: index[declaration.name].some is not a function
    at ResolveIndex.<anonymous> (d:\Projects\typescript-hero\out\src\caches\ResolveIndex.js:342:50)
    at next (<anonymous>)
    at __awaiter (d:\Projects\typescript-hero\out\src\caches\ResolveIndex.js:19:65)
    at __awaiter (d:\Projects\typescript-hero\out\src\caches\ResolveIndex.js:15:12)
    at ResolveIndex.createIndex (d:\Projects\typescript-hero\out\src\caches\ResolveIndex.js:320:16)
```

Actual error comes from line 390 at src\caches\ResolveIndex.ts.
Here is the latest state of variables just before crash (in JSON):

`from`:
```json
"ip"
```

`declaration`:
```json
"{"start":896,
"end":979,
"name":"toString",
"type":"string",
"isExported":true,
"parameters":[
{"start":921,"end":935,"name":"ip","type":"NodeBuffer"},
{"start":937,"end":952,"name":"offset","type":"number"},
{"start":954,"end":969,"name":"length","type":"number"}
],
"variables":[]}"
```

`key`:
```json
"ip"
```

`resource`:
```json
"{"start":523,"end":3630,"imports":[],"declarations":[{"start":612,"end":671,"name":"isEqual","type":"boolean","isExported":true,"parameters":[{"start":636,"end":647,"name":"ip1","type":"string"},{"start":649,"end":660,"name":"ip2","type":"string"}],"variables":[]},{"start":742,"end":825,"name":"toBuffer","type":"NodeBuffer","isExported":true,"parameters":[{"start":767,"end":777,"name":"ip","type":"string"},{"start":779,"end":794,"name":"buffer","type":"number"},{"start":796,"end":811,"name":"offset","type":"number"}],"variables":[]},{"start":896,"end":979,"name":"toString","type":"string","isExported":true,"parameters":[{"start":921,"end":935,"name":"ip","type":"NodeBuffer"},{"start":937,"end":952,"name":"offset","type":"number"},{"start":954,"end":969,"name":"length","type":"number"}],"variables":[]},{"start":1199,"end":1275,"name":"fromPrefixLen","type":"string","isExported":true,"parameters":[{"start":1229,"end":1249,"name":"prefixLength","type":"number"},{"start":1251,"end":1265,"name":"family","type":"string"}],"variables":[]},{"start":1380,"end":1435,"name":"mask","type":"string","isExported":true,"parameters":[{"start":1401,"end":1411,"name":"ip","type":"string"},{"start":1413,"end":1425,"name":"mask","type":"string"}],"variables":[]},{"start":1537,"end":1580,"name":"cidr","type":"string","isExported":true,"parameters":[{"start":1558,"end":1570,"name":"cidr","type":"string"}],"variables":[]},{"start":1690,"end":1730,"name":"not","type":"string","isExported":true,"parameters":[{"start":1710,"end":1720,"name":"ip","type":"string"}],"variables":[]},{"start":1847,"end":1899,"name":"or","type":"string","isExported":true,"parameters":[{"start":1866,"end":1876,"name":"ip","type":"string"},{"start":1878,"end":1889,"name":"mask","type":"string"}],"variables":[]},{"start":1992,"end":2039,"name":"isPrivate","type":"boolean","isExported":true,"parameters":[{"start":2018,"end":2028,"name":"ip","type":"string"}],"variables":[]},{"start":2131,"end":2177,"name":"isPublic","type":"boolean","isExported":true,"parameters":[{"start":2156,"end":2166,"name":"ip","type":"string"}],"variables":[]},{"start":2255,"end":2303,"name":"isLoopback","type":"boolean","isExported":true,"parameters":[{"start":2282,"end":2292,"name":"ip","type":"string"}],"variables":[]},{"start":2472,"end":2522,"name":"loopback","type":"string","isExported":true,"parameters":[{"start":2497,"end":2512,"name":"family","type":"string"}],"variables":[]},{"start":2921,"end":2984,"name":"address","type":"string","isExported":true,"parameters":[{"start":2945,"end":2958,"name":"name","type":"string"},{"start":2960,"end":2975,"name":"family","type":"string"}],"variables":[]},{"start":3090,"end":3133,"name":"toLong","type":"number","isExported":true,"parameters":[{"start":3113,"end":3123,"name":"ip","type":"string"}],"variables":[]},{"start":3240,"end":3285,"name":"fromLong","type":"string","isExported":true,"parameters":[{"start":3265,"end":3275,"name":"ip","type":"number"}],"variables":[]},{"start":3415,"end":3478,"name":"subnet","type":"SubnetInfo","isExported":true,"parameters":[{"start":3438,"end":3448,"name":"ip","type":"string"},{"start":3450,"end":3464,"name":"subnet","type":"string"}],"variables":[]},{"start":3574,"end":3627,"name":"cidrSubnet","type":"SubnetInfo","isExported":true,"parameters":[{"start":3601,"end":3613,"name":"cidr","type":"string"}],"variables":[]}],"exports":[],"resources":[],"usages":["ip1","ip2","ip","buffer","offset","NodeBuffer","length",null,"family","prefixLength","mask","cidr","name","subne...
```

`index` contains circular structure so is not possible to convert to JSON but obviously it does not contain `declaration.name` which is 'toString' in this context.

**Comment by buehler** [comment]
@gund 
Thank you very much! I don't know if I'd ever thought about that... !

Happy new year!
